Linux/Mac support and other fixes

rrooij opened the pull request:
# Fixed

* Rename the binary from xdelta to xdelta3
* Fix too many open file descriptors in the importing and writing of scenario.dat, which caused the patching to fail
* Replace the animated gif loading with a static hourglass icon to prevent Mono from crashing

# Patches in GNU/Linux and Mac OS X

Patches now working:
- btl.svo
- chat.svo
- effect.svo
- menu.svo
- scenario.dat: too many open files (Fixed)
- UI.svo
- PARAM.SFO
- TROPHY.TRP

Needs testing:
- chara.svo (fails at checksum)
- string.svo (my rip is bad)
- eboot.bin: does the generated eboot.bin from makeself actually work?

# TODO (if I have more time)

Some things that might be nice if I have more time:
* [DONE] Let the program stop complaining that ebootmod can't be found if it is installed anyway and in PATH.
* [DONE] Find a cross platform solution for ebootmod
* [DONE] Document prerequisites for Linux in the README

AdmiralCurtiss commented on ElfPatcher.cs:
```diff
@@ -8,8 +8,17 @@
+			OperatingSystem os = Environment.OSVersion;
+			PlatformID pid = os.Platform;
+			string patcherPath = "";
+			if (pid == PlatformID.Unix || pid == PlatformID.MacOSX) {
```
It's probably a better idea to check this the other way around. Most OSes should be able to use unself, it's just Windows we special-case.

AdmiralCurtiss commented on ElfPatcher.cs:
```diff
@@ -22,16 +31,20 @@ class ElfPatcher {
 				try {
-					RunEbootModAndKill( ebootModPath, "\"" + ebootPath + "\"" );
+					if (linuxOrMac) {
+						RunElfPatcherAndKill( patcherPath, " " + ebootPath + " " + elfPath);
```
`RunElfPatcherAndKill` is some real hacky code that runs ebootmod and then kills it mid-execution to get the decrypted eboot, you should probably call a regular program execution here instead.

AdmiralCurtiss commented on PatchForm.cs:
```diff
@@ -57,13 +57,26 @@ public partial class PatchForm : Form {
+					String.Format(warningMessage, "ebootMOD", ebootModPath),
 					"ebootMOD not found!", MessageBoxButtons.OK, MessageBoxIcon.Warning
 				);
+			} else if (!File.Exists (makeSelfPath)) {
```
You probably just want two cases here: One for Windows that checks for ebootmod, and one for non-Windows that checks for makeself and unself. Wouldn't want to extract makeself, restart the tool, and get a message that it *also* needs unself if it could just say that immediately, y'know?

rrooij replied:
By the way, what about using the makeself and unself executables on Windows too? We could set an environment variable to a specific folder to solve the confusion about the key folder. It would almost eradicate all platform specific code. See https://github.com/tenshou/ps3fail0verflow/blob/master/tools.c#L390 regarding the path of the keys.

rrooij commented:
Thanks for your feedback. I will work on it when I find time this week.

AdmiralCurtiss commented on ElfPatcher.cs:
```diff
@@ -22,16 +26,21 @@ class ElfPatcher {
+						Util.RunProgram(patcherPath, ebootPath + " " + elfPath, false, false);
+					}
+					else {
+						RunElfPatcherAndKill( patcherPath, "\"" + ebootPath + " " + elfPath );
```
You changed the Windows command line call here, it should be just `RunEbootModAndKill( ebootModPath, "\"" + ebootPath + "\"" );`

AdmiralCurtiss commented on ElfPatcher.cs:
```diff
@@ -22,16 +26,21 @@ class ElfPatcher {
-					RunEbootModAndKill( ebootModPath, "\"" + ebootPath + "\"" );
+					if (!Util.isRunningOnWindows()) {
+						Console.WriteLine(patcherPath + " " + ebootPath + " " + elfPath);
+						Util.RunProgram(patcherPath, ebootPath + " " + elfPath, false, false);
```
You may need to quote the two filenames here so this works if the path contains spaces.

AdmiralCurtiss commented:
Oh and I have no problems with allowing `unself` on Windows too if we have a foolproof way to let that detect the keys, though I would like to keep the `ebootmod` support as well just in case.

rrooij commented:
Thanks for your feedback.

A few things I'd like to let you know before you decide to merge:

* Checksum of char fails, it's the only thing that can't be patched on Linux and Mac. Maybe this should be covered in the README or fixed before merging this. I don't know whether the resulting archive is still valid though. I have to test this on a real PS3. If only the checksum is different, but it works correctly, maybe checking for two checksums (one for Windows and one for others) would be the way to solve this.
* I haven't tested it on Windows yet. This is certainly a requirement before merging it to master
* Patching the EBOOT.bin is going okay according to the tool, but I haven't tested it on a real PS3 yet.

I understand your concern regarding dropping support for ebootmod entirely.

AdmiralCurtiss commented:
I tested this on Windows just now, works fine. I'll also patch on Linux and check if the modified EBOOT works before merging, we can fix the chara.svo failure in another issue.

rrooij commented:
Glad to hear that it's still working fine on Windows.

And thanks for trying it out on Linux too! If the eboot fails, it would be because of the signing parameters I think.

AdmiralCurtiss commented:
Alright, a few small things I noticed:

- Error message still refer to Windows-like names, eg it complains that it can't find 'comptoe.exe' and similar when you open the patcher without the relevant executables in the right location, eboot failure says that ebootMOD failed, that kinda stuff. Error message for xdelta3 also still says that xdelta failed, without the 3.
- Also, the program check on open checks for `xdelta` instead of `xdelta3`. Maybe we should remove this check altogether though, since it's quite likely that `xdelta3` will be an installed program in PATH on non-Windows rather than in the same folder as the patcher.
- I got a bit confused by the requested folder for makeself/unself being `failOverflowTools` with an O when the repo I cloned them from is called `fail0verflow` with a zero, maybe those should be consistent. Or just call the folder something simpler.

Haven't tested the eboot on console yet either, but this seems fine otherwise.

AdmiralCurtiss commented:
Also, I'd appreciate it if you could squash your fixups into the relevant commits before merging.

rrooij commented:
Thanks for testing this. I will fix those things and squash the commits.

rrooij commented:
I've pushed my rebased commits here in a separate branch, just to be sure:
https://github.com/rrooij/ToVPatcher/commits/rebase_test

If you're okay with the rebases, I will merge them to this branch to make it easier for you to merge.

AdmiralCurtiss commented:
That rebase looks fine to me.

AdmiralCurtiss commented:
Comparing the logs, it looks like `YUR_C201.DAT` is getting generated differently on Linux, which in turn affects the checksum of `chara.svo`. I'll check this out, probably not too complicated to fix.

AdmiralCurtiss commented:
The Linux-patched EBOOT.BIN does not work, at least on Rebug 4.81 using webMAN to launch the game from internal HDD. Error code 80010007 during game launch. The Windows-patched version loads fine under the same setup.

rrooij commented:
Hmmmmm.... It's probably a wrong argument when calling makeself. I'm not sure though what could be the cause.

I recently gave my modded PS3 away as gift, so I can't test it anymore unfortunately.

See: http://www.psdevwiki.com/ps3/Dev_Tools#fail0VERFLOW for possible makeself parameters. My guess is that keytype should be changed from app to iso, but I'm not sure and not an expert on this issue.

rrooij commented:
We could also use [self_rebuilder](https://github.com/euss/ps3tools/blob/master/self_rebuilder.c). It only requires <input.elf> <output.self> <original.self>

rrooij commented:
I fucked up the rebase a bit now by the way. Be sure not to merge anything yet. I'm also experimenting with self_rebuilder now.

AdmiralCurtiss commented:
Err, just replace this branch with your rebase_test one. Try `git push origin rebase_test:linux_support`, maybe with a `--force`.

AdmiralCurtiss commented:
Figured out why chara.svo fails, see #6.

rrooij commented:
I've implemented self_rebuilder instead of makeself. It patches correctly, but I don't know whether it works on a real PS3. Would you mind testing it? You can find a good copy of a repo with self_rebuilder here:

https://github.com/Evilnat/ps3tools-master_3.XX-4.XX

AdmiralCurtiss commented on README.md:
```diff
@@ -78,10 +78,10 @@ Instructions
-  The failOverflow tools to de- and encrypt are found [here](https://github.com/eltair/fail0verflow). Create a folder called "eboot_tools"
-  in the path of the ToVPatcher.exe binary and copy the makeself and unself binaries there.
+  The failOverflow tools to de- and encrypt are found [here](https://github.com/Evilnat/ps3tools-master_3.XX-4.XX). Create a folder called "eboot_tools"
+  in the path of the ToVPatcher.exe binary and copy the self\_rebuilder and unself binaries there.
```
Typo in self_rebuilder here.

rrooij replied:
Actually, it's not a typo but an escape sequence. Otherwise Markdown will recognize the text as italics.

See: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#emphasis

AdmiralCurtiss replied:
Aaah, I see. I'm surprised Markdown parses that as italics in the middle of a word.

rrooij replied:
It was actually my editor (Emacs) that displayed the text as italics starting from "_tools" to "self_. GitHub's Markdown parsing is a little bit different and doesn't recognize that piece of text as italics.

So I will pick your preference. Feel free to choose what you prefer and I will rebase it.

AdmiralCurtiss commented:
Bad news, this doesn't work either, same error.

I did some digging to figure out how ebootMOD generates the file, and from what I can tell it uses some variant of make_self from ps3publictools. However, none of the variants of that I can find online quite match the Windows output and trying to run the generated EBOOTs from the code I can find just seem to crash the console on game boot.

Why is PS3 homebrew such a mess?

rrooij commented:
Ah that's bad news. Maybe a stupid question, but are you sure you have Cobra enabled and the like?

I don't know either what causes the different output.  Maybe you can use readself on both generated EBOOTs and diff them.

AdmiralCurtiss commented:
Yeah, Cobra is enabled.

I compared the readself output of the EBOOTs and the ebootmod-patched one differs pretty significantly from the original, while the linux-patched ones only differ in a few lines of the 'Keys'. And yet only the ebootmod-patched one works. I have no idea how to interpret this though, sorry.

rrooij commented:
Hmmm.. Maybe there are some other people from the scene that could lend us a bit of help. Thanks for comparing the output!

rrooij commented:
I made a thread here.

http://www.psx-place.com/threads/need-help-porting-tales-of-vesperia-patcher-to-linux.16652/

Someone suggested just making a xdelta for the whole EBOOT.

rrooij commented:
Would you mind testing this? This is just a test, if it works I will make the commit prettier and update the instructions in README.

For now, just clone: https://github.com/HACKERCHANNEL/PS3Py in the binary folder and be sure that fself.py is executable.

The downside of this is that it can run only on DEX or REX firmware as far as I know, since it is a debug eboot. But if it works, it is better than nothing and we could look for other solutions in the future.

AdmiralCurtiss commented:
That actually does boot on my console. I think that's a decent compromise.

It's probably better to execute `python` with the script as the argument, I'm guessing.

rrooij commented:
Awesome! I will rebase and clean my stuff up tomorrow.

rrooij commented:
And executing with Python as argument can be tricky. In some systems it refers to Python2 and in some Python3.

rrooij commented:
Did my changes and tested it. I think it is OK now.

The only thing I'd like to change is porting the PS3Py tools to Python3. Python2 will stop receiving security updates in the not so long future. https://pythonclock.org/

But this is for the long run and maybe we will get makeself working before that happens.

AdmiralCurtiss commented on PatchForm.cs:
```diff
@@ -57,13 +57,21 @@ public partial class PatchForm : Form {
-					"ebootMOD could not be found at " + ebootModPath + "." + Environment.NewLine +
-					"ebootMOD is required to patch EBOOT.BIN. Please read the readme, find a copy of ebootMOD, and place it at the appropriate location." + Environment.NewLine +
-					"The patcher will still run, but will not be able to patch EBOOT.BIN until you do so.",
+			string makeSelfPath = Path.GetFullPath( "failOverflowTools/makeself" );
```
This isn't right, we're now using fself.py.

AdmiralCurtiss commented on PatchForm.cs:
```diff
@@ -57,13 +57,21 @@ public partial class PatchForm : Form {
 				);
+			} else if ( (!File.Exists (makeSelfPath) || !File.Exists (unSelfPath) ) && !Util.isRunningOnWindows()) {
+				MessageBox.Show (this,
+					String.Format(warningMessage, "makeself and unself", makeSelfPath),
```
Here too.

AdmiralCurtiss commented on PatchForm.cs:
```diff
@@ -57,13 +57,21 @@ public partial class PatchForm : Form {
+			} else if ( (!File.Exists (makeSelfPath) || !File.Exists (unSelfPath) ) && !Util.isRunningOnWindows()) {
+				MessageBox.Show (this,
+					String.Format(warningMessage, "makeself and unself", makeSelfPath),
+					"makeSelf and unself not found!", MessageBoxButtons.OK, MessageBoxIcon.Warning
```
And here.

AdmiralCurtiss commented on ElfPatcher.cs:
```diff
@@ -10,7 +10,7 @@ namespace ToVPatcher {
 		public static void PatchElf( string ebootPath, string patchDir, string outDir, string outMd5 = null, BackgroundWorker worker = null ) {
 			
-			string patcherPath = "failOverflowTools/unself";
+			string patcherPath = "eboot_tools/unself";
```
Ah, I see. I'd prefer if you squashed those so that the commit that introduces the Linux patcher already uses the correct tool and names.

rrooij commented:
Done. I squashed the executable fixes and the fself and unself commit. The only thing that bothers me  a bit is that the xdelta3.exe fixes were also in the commit with the "Fix executables" commit message, which are now introduced in the makeself and fself commit because of the squash.

AdmiralCurtiss commented:
I mean, if you want you can separate them out, interactive rebase is pretty powerful, but you don't have to. Whichever you want, I'm fine with either.

rrooij commented:
I did it. It was less work than I thought indeed. No xdelta stuff can be found in the unself and fself commit.

AdmiralCurtiss commented:
Alright, merged it. Thank you!
